Add support for `Audio` and `Image` feature in `push_to_hub`

mariosasko opened the pull request:
Add support for the `Audio` and the `Image` feature in `push_to_hub`. 

The idea is to remove local path information and store file content under "bytes" in the Arrow table before the push.

My initial approach (https://github.com/huggingface/datasets/commit/34c652afeff9686b6b8bf4e703c84d2205d670aa) was to use a map transform similar to [`decode_nested_example`](https://github.com/huggingface/datasets/blob/5e0f6068741464f833ff1802e24ecc2064aaea9f/src/datasets/features/features.py#L1023-L1056) while having decoding turned off, but I wasn't satisfied with the code quality, so I ended up using the `temporary_assignment` decorator to override `cast_storage`, which allows me to directly modify the underlying storage (the final op is similar to `Dataset.cast`) and results in a much simpler code. 

Additionally, I added the `allow_cast` flag that can disable this behavior in the situations where it's not needed (e.g. the dataset is already in the correct format for the Hub, etc.)

lhoestq reviewed (commented):
Cool thanks !

Also cc @patrickvonplaten @anton-l it means that when calling push_to_hub, the audio bytes are embedded in the parquet files (we don't upload the audio files themselves)

patrickvonplaten commented:
> Cool thanks !
> 
> Also cc @patrickvonplaten @anton-l it means that when calling push_to_hub, the audio bytes are embedded in the parquet files (we don't upload the audio files themselves)

Just to verify quickly the size of the dataset doesn't change in this case no? E.g. if a dataset has say 20GB in size when stored in `.mp3` format it could have up to 100GB when stored in WAV. But since we are just taking the bytes here a 20GB .mp3 dataset would also have 20GB when stored in parquet no?

mariosasko commented:
@lhoestq  I've addressed your comments. Additionally, I've modified `cast_storage` to account for possible null (`None`) values.

@patrickvonplaten Yes, the dataset size stays the same (at least because Parquet files are compressed).

lhoestq commented on src/datasets/arrow_dataset.py:
```diff
@@ -3577,6 +3619,7 @@ def push_to_hub(
         token: Optional[str] = None,
         branch: Optional[str] = None,
         shard_size: Optional[int] = 500 << 20,
+        allow_cast: bool = True,
```
We can name this parameter embed_external_files to keep things simple if you want

lhoestq commented on src/datasets/arrow_dataset.py:
```diff
@@ -3512,14 +3530,65 @@ def _push_parquet_shards_to_hub(
+                    for shard in shards:
+                        format = shard.format
+                        shard = shard.with_format("arrow")
+                        shard = shard.map(
```
It would be nice to disable this tqdm bar, this can be done in a subsequent PR

lhoestq reviewed (approved):
Awesome thank you !

The `_push_parquet_shards_to_hub` method starts to be quite long, feel free to add some comments to explain the different steps. You can also add some info/debug logging where it can make sense.

lhoestq commented on src/datasets/table.py:
```diff
@@ -1123,3 +1123,40 @@ def table_cast(table: pa.Table, schema: pa.Schema):
+    features = Features.from_arrow_schema(table.schema)
+
+    def _visit(array, feature):
+        if isinstance(array, pa.ChunkedArray):
+            for chunk in array.chunks:
+                _visit(chunk, feature)
```
You can also do `array.nbytes` too on a ChunkedArray no ?

mariosasko replied:
Yes, but we call `nbytes` on the struct arrays with image data, and these arrays are possibly nested inside a ChunkedArray.

lhoestq replied:
Ok I see !

lhoestq commented:
Feel free to merge if it's all good to you :)
